README
NAME
    SHARYANTO::Proc::Util - OS-process-related routines

VERSION
    version 0.61

SYNOPSIS
FUNCTIONS
    None are exported by default, but they are exportable.

  get_parent_processes($pid[, \%opts]) => ARRAY
    Return an array containing information about parent processes of $pid up
    to the parent of all processes (usually "init"). If $pid is not
    mentioned, it defaults to $$. The immediate parent is in the first
    element of array, followed by its parent, and so on. For example:

     [{name=>"perl",pid=>13134}, {name=>"konsole",pid=>2232}, {name=>"init",pid=>1}]

    Currently retrieves information by calling pstree program. Return undef
    on failure.

    Known options:

    *   method => STR (default: "proctable")

        Either "proctable" (the default, which means to use
        Proc::ProcessTable) or "pstree" (which uses the pstree command,
        which might not be portable between Unices).

SEE ALSO
    SHARYANTO

    Proc::ProcessTable. Pros: does not depend on pstree command, process
    names not truncated by pstree. Cons: a little bit more heavyweight (uses
    File::Spec, Cwd, File::Find).

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/SHARYANTO-Proc-Util>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-SHARYANTO-Proc-Util>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=SHARYANTO-Proc-Util>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for SHARYANTO-Proc-Util

0.61    2014-04-14 (SHARYANTO)

        - Skip test if get_parent_processes() returns undef [CT].


0.60    2014-04-10 (SHARYANTO)

        - Trap failure in loading Proc::ProcessTable so this module can still be
          used without it.


0.59    2013-11-11 (SHARYANTO)

        - Use Proc::ProcessTable by default instead of 'pstree' command
          (suggested by SREZIC, [RT#91932]).


0.58    2013-11-11 (SHARYANTO)

        - No functional changes. Revise test script to avoid CT failure reports.


0.57    2013-09-26 (SHARYANTO)

        - Fix failure on Ubuntu because pstree output is cut at 132 columns,
          need to add -l (thanks Olivier Mengué).


0.56    2013-09-10 (SHARYANTO)

        - No functional changes. Split from SHARYANTO-Utils to minimize
          dependencies (I'm trying to make Text::ANSITable work under Windows).


lib/SHARYANTO/Proc/Util.pm
package SHARYANTO::Proc::Util;

use 5.010001;
use strict;
use warnings;

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(get_parent_processes);

our $VERSION = '0.61'; # VERSION

sub get_parent_processes {
    my ($pid, $opts) = @_;
    $pid //= $$;
    $opts //= {};

    my %proc;
    if (($opts->{method} // 'proctable') eq 'pstree') {
        # things will be simpler if we use the -s option, however not all
        # versions of pstree supports it. -l is for --long (to avoid pstree to
        # cut its output at 132 cols)

        my @lines = `pstree -pAl`;
        return undef unless @lines;

        my @p;
        for (@lines) {
            my $i = 0;
            while (/(?: (\s*(?:\|-?|`-)) | (.+?)\((\d+)\) )
                    (?: -[+-]- )?/gx) {
                unless ($1) {
                    my $p = {name=>$2, pid=>$3};
                    $p[$i] = $p;
                    $p->{ppid} = $p[$i-1]{pid} if $i > 0;
                    $proc{$3} = $p;
                }
                $i++;
            }
        }
        #use Data::Dump; dd \%proc;
    } else {
        eval { require Proc::ProcessTable };
        return undef if $@;

        state $pt = Proc::ProcessTable->new;
        for my $p (@{ $pt->table }) {
            $proc{ $p->{pid} } = {
                name=>$p->{fname}, pid=>$p->{pid}, ppid=>$p->{ppid},
            };
        }
    }

    my @p = ();
    my $cur_pid = $pid;
    while (1) {
        return if !$proc{$cur_pid};
        $proc{$cur_pid}{name} = $1 if $proc{$cur_pid}{name} =~ /\A\{(.+)\}\z/;
        push @p, $proc{$cur_pid};
        $cur_pid = $proc{$cur_pid}{ppid};
        last unless $cur_pid;
    }
    shift @p; # delete cur process

    \@p;
}

# ABSTRACT: OS-process-related routines

__END__

=pod

=encoding UTF-8

=head1 NAME

SHARYANTO::Proc::Util - OS-process-related routines

=head1 VERSION

version 0.61

=head1 SYNOPSIS

=head1 FUNCTIONS

None are exported by default, but they are exportable.

=head2 get_parent_processes($pid[, \%opts]) => ARRAY

Return an array containing information about parent processes of C<$pid> up to
the parent of all processes (usually C<init>). If C<$pid> is not mentioned, it
defaults to C<$$>. The immediate parent is in the first element of array,
followed by its parent, and so on. For example:

 [{name=>"perl",pid=>13134}, {name=>"konsole",pid=>2232}, {name=>"init",pid=>1}]

Currently retrieves information by calling B<pstree> program. Return undef on
failure.

Known options:

=over

=item * method => STR (default: C<proctable>)

Either C<proctable> (the default, which means to use L<Proc::ProcessTable>) or
C<pstree> (which uses the B<pstree> command, which might not be portable between
Unices).

=back

=head1 SEE ALSO

L<SHARYANTO>

L<Proc::ProcessTable>. Pros: does not depend on pstree command, process names
not truncated by pstree. Cons: a little bit more heavyweight (uses File::Spec,
Cwd, File::Find).

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/SHARYANTO-Proc-Util>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-SHARYANTO-Proc-Util>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=SHARYANTO-Proc-Util>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "OS-process-related routines",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "SHARYANTO-Proc-Util",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "SHARYANTO::Proc::Util",
  "PREREQ_PM" => {
    "Proc::ProcessTable" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "File::Which" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.61",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "File::Spec" => 0,
  "File::Which" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Proc::ProcessTable" => 0,
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



